File: 22/solve.cpp
#include <iostream>
#include <vector>

int main()
{
    std::vector<long> input;
    while (true) {
        long num;
        std::cin >> num;

        if (std::cin.eof())
            break;

        input.push_back(num);
    }

    long a {};
    long b {};

    std::vector<long> scores(19 * 19 * 19 * 19);
    std::vector<long> seen(19 * 19 * 19 * 19, -1);

    std::vector<long> suffix;
    for (long i = 0; i < input.size(); ++i) {
        long secret = input[i];
        long last = secret % 10;

        suffix.clear();

        for (int k = 0; k < 2000; ++k) {
            secret ^= secret << 6;
            secret &= 0xffffff;

            secret ^= secret >> 5;
            secret &= 0xffffff;

            secret ^= secret << 11;
            secret &= 0xffffff;

            long digit = secret % 10;
            long delta = digit - last;
            last = digit;

            suffix.push_back(delta + 9);
            if (suffix.size() > 4) {
                suffix.erase(suffix.begin());
            } else {
                continue;
            }

            long idx = 6859 * suffix[0] + 361 * suffix[1] + 19 * suffix[2] + suffix[3];
            if (seen[idx] < i) {
                scores[idx] += digit;
                seen[idx] = i;
                b = std::max(b, scores[idx]);
            }
        }

        a += secret;
    }

    std::cout << a << "\n";
    std::cout << b << "\n";
}

File: 22/solve.py
import collections


def solve_p1(path):
    with open(path) as f:
        nums = [int(x) for x in f.read().splitlines()]

    res = 0
    for init in nums:
        secret = init
        for _ in range(2000):
            secret ^= secret << 6
            secret &= 0xffffff

            secret ^= secret >> 5
            secret &= 0xffffff

            secret ^= secret << 11
            secret &= 0xffffff
        res += secret

    return res


def solve_p2(path):
    with open(path) as f:
        nums = [int(x) for x in f.read().splitlines()]

    res = 0
    scores = collections.Counter()
    for init in nums:
        secret = init
        suffix = tuple()
        last = init % 10
        buyer_score = collections.Counter()

        for _ in range(2000):
            secret ^= secret << 6
            secret &= 0xffffff

            secret ^= secret >> 5
            secret &= 0xffffff

            secret ^= secret << 11
            secret &= 0xffffff
            digit = secret % 10

            delta = digit - last
            last = digit

            suffix = (*suffix, delta)
            if len(suffix) > 4:
                suffix = suffix[1:]
            else:
                continue

            if suffix not in buyer_score:
                buyer_score[suffix] = digit

        scores += buyer_score

    res = max(scores.values())

    return res


def main():
    assert solve_p1("example") == 37327623
    print(solve_p1("input"))

    assert solve_p2("example2") == 23
    print(solve_p2("input"))


if __name__ == "__main__":
    main()
